File: sdk/hypercli/renders.py
from __future__ import annotations

"""Renders API"""
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, List

if TYPE_CHECKING:
    from .http import HTTPClient
from .http import APIError


@dataclass
class Render:
    render_id: str
    state: str
    template: str | None = None
    render_type: str | None = None
    tags: list[str] | None = None
    result_url: str | None = None
    error: str | None = None
    created_at: float | None = None
    started_at: float | None = None
    completed_at: float | None = None

    @classmethod
    def from_dict(cls, data: dict) -> "Render":
        return cls(
            render_id=data.get("id") or data.get("render_id", ""),
            state=data.get("state", ""),
            template=data.get("template") or data.get("meta", {}).get("template"),
            render_type=data.get("type") or data.get("render_type"),
            tags=list(data.get("tags") or []),
            result_url=data.get("result_url"),
            error=data.get("error"),
            created_at=data.get("created_at"),
            started_at=data.get("started_at"),
            completed_at=data.get("completed_at"),
        )


@dataclass
class RenderStatus:
    render_id: str
    state: str
    progress: float | None = None

    @classmethod
    def from_dict(cls, data: dict) -> "RenderStatus":
        return cls(
            render_id=data.get("id") or data.get("render_id", ""),
            state=data.get("state", ""),
            progress=data.get("progress"),
        )


class Renders:
    """Renders API wrapper"""

    def __init__(self, http: "HTTPClient", auth_http: "HTTPClient" | None = None):
        self._http = http
        self._auth_http = auth_http or http
        self._auth_me_cache: dict[str, Any] | None = None

    def _auth_me(self) -> dict[str, Any]:
        if self._auth_me_cache is None:
            self._auth_me_cache = self._auth_http.get("/auth/me")
        return self._auth_me_cache

    def _supports_subscription_family(self, family: str, resource: str | None = None) -> bool:
        try:
            auth_me = self._auth_me()
        except APIError:
            return False
        if not auth_me.get("has_active_subscription"):
            return False
        if auth_me.get("auth_type") == "user":
            return True
        capabilities = set(auth_me.get("capabilities") or [])
        if f"{family}:*" in capabilities:
            return True
        if resource and f"{family}:{resource}" in capabilities:
            return True
        return False

    def _post_flow(self, flow_type: str, payload: dict[str, Any]) -> dict[str, Any]:
        primary = f"/agents/flow/{flow_type}" if self._supports_subscription_family("renders", flow_type) else f"/api/flow/{flow_type}"
        try:
            return self._http.post(primary, json=payload)
        except APIError as exc:
            if primary.startswith("/agents/flow/") and exc.status_code in {403, 404}:
                return self._http.post(f"/api/flow/{flow_type}", json=payload)
            raise

    def _get_render(self, render_id: str, *, status: bool = False) -> dict[str, Any]:
        suffix = "/status" if status else ""
        primary = f"/agents/flow/renders/{render_id}{suffix}" if self._supports_subscription_family("renders") else f"/api/renders/{render_id}{suffix}"
        try:
            return self._http.get(primary)
        except APIError as exc:
            if primary.startswith("/agents/flow/") and exc.status_code in {403, 404}:
                return self._http.get(f"/api/renders/{render_id}{suffix}")
            raise

    def _delete_render(self, render_id: str) -> dict:
        primary = f"/agents/flow/renders/{render_id}" if self._supports_subscription_family("renders") else f"/api/renders/{render_id}"
        try:
            return self._http.delete(primary)
        except APIError as exc:
            if primary.startswith("/agents/flow/") and exc.status_code in {403, 404}:
                return self._http.delete(f"/api/renders/{render_id}")
            raise

    def list(
        self,
        state: str = None,
        template: str = None,
        type: str = None,
        tags: list[str] | None = None,
    ) -> list[Render]:
        """List all renders.

        Args:
            state: Filter by state (e.g., "pending", "running", "completed")
            template: Filter by template name
            type: Filter by render type (e.g., "comfyui")
        """
        params = {}
        if state:
            params["state"] = state
        if template:
            params["template"] = template
        if type:
            params["type"] = type
        if tags:
            params["tag"] = list(tags)

        data = self._http.get("/api/renders", params=params or None)
        # Handle paginated response
        items = data.get("items", data) if isinstance(data, dict) else data
        return [Render.from_dict(r) for r in items]

    def get(self, render_id: str) -> Render:
        """Get render details"""
        data = self._get_render(render_id)
        return Render.from_dict(data)

    def create(
        self,
        params: dict,
        render_type: str = "comfyui",
        notify_url: str = None,
        tags: list[str] | None = None,
    ) -> Render:
        """Create a new render.

        Args:
            params: Render parameters (workflow-specific)
            render_type: Type of render (default: "comfyui")
            notify_url: Optional webhook URL for completion notification
        """
        payload = {
            "type": render_type,
            "params": params,
        }
        if notify_url:
            payload["notify_url"] = notify_url
        if tags:
            payload["tags"] = list(tags)

        data = self._http.post("/api/renders", json=payload)
        return Render.from_dict(data)

    def cancel(self, render_id: str) -> dict:
        """Cancel a render"""
        return self._delete_render(render_id)

    def status(self, render_id: str) -> RenderStatus:
        """Get render status (lightweight polling endpoint)"""
        data = self._get_render(render_id, status=True)
        return RenderStatus.from_dict(data)

    # =========================================================================
    # Flow endpoints - simplified interfaces
    # =========================================================================

    def create_flow(self, flow_type: str, **kwargs) -> Render:
        """Create a flow render via subscription flow when available, otherwise paid flow."""
        payload = {k: v for k, v in kwargs.items() if v is not None}
        data = self._post_flow(flow_type, payload)
        return Render.from_dict(data)

    def text_to_image(
        self,
        prompt: str,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate an image using Qwen-Image (great for text in images).

        Args:
            prompt: Text description of the image
            negative: Optional negative prompt (things to avoid)
            width: Optional output width
            height: Optional output height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.text_to_image("a cat wearing sunglasses")
        """
        return self.create_flow("text-to-image", prompt=prompt, negative=negative, width=width, height=height, notify_url=notify_url)

    def text_to_image_hidream(
        self,
        prompt: str,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate an image using HiDream I1 Full (highest quality).

        Args:
            prompt: Text description of the image
            negative: Optional negative prompt (things to avoid)
            width: Optional output width
            height: Optional output height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.text_to_image_hidream("a mystical forest")
        """
        return self.create_flow("text-to-image-hidream", prompt=prompt, negative=negative, width=width, height=height, notify_url=notify_url)

    def text_to_video(
        self,
        prompt: str,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate a video using Wan 2.2 14B.

        Args:
            prompt: Text description of the video
            negative: Optional negative prompt (things to avoid)
            width: Optional video width
            height: Optional video height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.text_to_video("a cat walking through a garden")
        """
        return self.create_flow("text-to-video", prompt=prompt, negative=negative, width=width, height=height, notify_url=notify_url)

    def image_to_video(
        self,
        prompt: str,
        image_url: str = None,
        file_ids: List[str] = None,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Animate an image using Wan 2.2 Animate.

        Args:
            prompt: Description of the motion/animation
            image_url: URL of the image to animate
            file_ids: Pre-uploaded file IDs [image_id]
            negative: Optional negative prompt (things to avoid)
            width: Optional video width
            height: Optional video height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.image_to_video("dancing", image_url="https://example.com/img.png")
        """
        return self.create_flow("image-to-video", prompt=prompt, image_url=image_url, file_ids=file_ids, negative=negative, width=width, height=height, notify_url=notify_url)

    def speaking_video(
        self,
        prompt: str,
        image_url: str = None,
        audio_url: str = None,
        file_ids: List[str] = None,
        negative: str = None,
        length: int = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate a lip-sync video using HuMo.

        Provide inputs as URLs or pre-uploaded file IDs:
        - URLs: image_url + audio_url
        - File IDs: file_ids=[image_file_id, audio_file_id]

        Args:
            prompt: Description of the scene/character
            image_url: URL of the face/character image
            audio_url: URL of the audio/speech file
            file_ids: Pre-uploaded file IDs [image_id, audio_id]
            negative: Optional negative prompt (things to avoid)
            length: Video length in frames at 25fps (calculated from audio duration)
            width: Optional video width
            height: Optional video height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.speaking_video(
                "A person talking to camera",
                image_url="https://example.com/face.png",
                audio_url="https://example.com/speech.mp3"
            )
        """
        return self.create_flow("speaking-video", prompt=prompt, image_url=image_url, audio_url=audio_url, file_ids=file_ids, negative=negative, length=length, width=width, height=height, notify_url=notify_url)

    def speaking_video_wan(
        self,
        prompt: str,
        image_url: str,
        audio_url: str,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate an audio-driven video using Wan 2.2 S2V.

        Args:
            prompt: Description of the scene/action
            image_url: URL of the image
            audio_url: URL of the audio file
            negative: Optional negative prompt (things to avoid)
            width: Optional video width
            height: Optional video height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.speaking_video_wan(
                "The person is singing",
                "https://example.com/face.png",
                "https://example.com/song.mp3"
            )
        """
        return self.create_flow("speaking-video-wan", prompt=prompt, image_url=image_url, audio_url=audio_url, negative=negative, width=width, height=height, notify_url=notify_url)

    def image_to_image(
        self,
        prompt: str,
        image_urls: List[str] = None,
        file_ids: List[str] = None,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Transform images using Qwen Image Edit with 1-3 input images.

        Args:
            prompt: Description of the transformation
            image_urls: List of 1-3 image URLs (first is main, others are references)
            file_ids: Pre-uploaded file IDs [main_id, ref1_id, ref2_id]
            negative: Optional negative prompt (things to avoid)
            width: Optional output width
            height: Optional output height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.image_to_image(
                "Apply the artistic style from the references",
                image_urls=["https://example.com/subject.jpg", "https://example.com/style.jpg"]
            )
        """
        return self.create_flow("image-to-image", prompt=prompt, image_urls=image_urls, file_ids=file_ids, negative=negative, width=width, height=height, notify_url=notify_url)

    def first_last_frame_video(
        self,
        prompt: str,
        start_image_url: str = None,
        end_image_url: str = None,
        file_ids: List[str] = None,
        negative: str = None,
        width: int = None,
        height: int = None,
        notify_url: str = None,
    ) -> Render:
        """Generate video morphing between two images using Wan 2.2.

        Args:
            prompt: Description of the transition/motion
            start_image_url: URL of the starting frame
            end_image_url: URL of the ending frame
            file_ids: Pre-uploaded file IDs [start_image_id, end_image_id]
            negative: Optional negative prompt (things to avoid)
            width: Optional video width
            height: Optional video height
            notify_url: Optional webhook URL for completion notification

        Example:
            render = client.renders.first_last_frame_video(
                "smooth transition from day to night",
                start_image_url="https://example.com/day.png",
                end_image_url="https://example.com/night.png"
            )
        """
        return self.create_flow("first-last-frame-video", prompt=prompt, start_image_url=start_image_url, end_image_url=end_image_url, file_ids=file_ids, negative=negative, width=width, height=height, notify_url=notify_url)

    def audio_to_text(
        self,
        audio_url: str = None,
        file_ids: List[str] = None,
        notify_url: str = None,
    ) -> Render:
        """Transcribe audio/video to text using WhisperX.

        Provide input as a URL or pre-uploaded file ID:
        - URL: audio_url="https://..."
        - File ID: file_ids=["audio_file_id"]

        Args:
            audio_url: URL of the audio or video file to transcribe
            file_ids: Pre-uploaded file IDs [audio_id]
            notify_url: Optional webhook URL for completion notification

        Returns:
            Render object. When completed, result_url points to a JSON file
            containing {"text": "transcription..."}.

        Example:
            render = client.renders.audio_to_text("https://example.com/recording.mp3")
            render = client.renders.audio_to_text(file_ids=["abc123"])
        """
        return self._flow("/api/flow/audio-to-text", audio_url=audio_url, file_ids=file_ids, notify_url=notify_url)

    def text_to_speech(
        self,
        text: str,
        mode: str = "custom",
        language: str = "Auto",
        speaker: str = None,
        style: str = None,
        model_size: str = None,
        voice_description: str = None,
        ref_audio_url: str = None,
        file_ids: List[str] = None,
        ref_text: str = None,
        use_xvector_only: bool = None,
        notify_url: str = None,
    ) -> Render:
        """Generate speech from text using Qwen3-TTS.

        Three modes:
        - "custom": Predefined speakers with optional style instructions
        - "design": Describe any voice in natural language
        - "clone": Clone a voice from reference audio

        Args:
            text: Text to synthesize
            mode: TTS mode ("custom", "design", or "clone")
            language: Language (Auto, English, Chinese, Japanese, Korean, etc.)
            speaker: Speaker name for custom mode (Ryan, Serena, etc.)
            style: Style instruction for custom mode (e.g. "Speak cheerfully")
            model_size: Model size ("0.6B" or "1.7B")
            voice_description: Voice description for design mode
            ref_audio_url: Reference audio URL for clone mode
            ref_text: Transcript of reference audio for clone mode
            use_xvector_only: Use x-vector only for clone mode
            notify_url: Optional webhook URL for completion notification

        Returns:
            Render object. When completed, result_url points to the WAV audio file.

        Example:
            render = client.renders.text_to_speech("Hello!", mode="design",
                voice_description="A young Indian male, enthusiastic")
        """
        return self._flow(
            "/api/flow/text-to-speech",
            text=text,
            mode=mode,
            language=language,
            speaker=speaker,
            style=style,
            model_size=model_size,
            voice_description=voice_description,
            ref_audio_url=ref_audio_url,
            file_ids=file_ids,
            ref_text=ref_text,
            use_xvector_only=use_xvector_only,
            notify_url=notify_url,
        )

File: ts-sdk/src/renders.ts
/**
 * Renders API - managed AI rendering workflows
 */
import type { HTTPClient } from './http.js';
import { APIError } from './errors.js';

export interface Render {
  renderId: string;
  state: string;
  template: string | null;
  renderType: string | null;
  tags: string[] | null;
  resultUrl: string | null;
  error: string | null;
  createdAt: number | null;
  startedAt: number | null;
  completedAt: number | null;
}

export interface RenderStatus {
  renderId: string;
  state: string;
  progress: number | null;
}

function renderFromDict(data: any): Render {
  return {
    renderId: data.id || data.render_id || '',
    state: data.state || '',
    template: data.template || data.meta?.template || null,
    renderType: data.type || data.render_type || null,
    tags: Array.isArray(data.tags) ? data.tags : null,
    resultUrl: data.result_url || null,
    error: data.error || null,
    createdAt: data.created_at || null,
    startedAt: data.started_at || null,
    completedAt: data.completed_at || null,
  };
}

function renderStatusFromDict(data: any): RenderStatus {
  return {
    renderId: data.id || data.render_id || '',
    state: data.state || '',
    progress: data.progress ?? null,
  };
}

export class Renders {
  static readonly DEFAULT_WAIT_TIMEOUT = 3600_000;
  static readonly DEFAULT_QUEUE_GRACE = 1800_000;
  static readonly DEFAULT_ACTIVE_GRACE = 300_000;

  private authMeCache: any | null = null;

  constructor(private http: HTTPClient, private authHttp: HTTPClient = http) {}

  private async authMe(): Promise<any> {
    if (this.authMeCache == null) {
      this.authMeCache = await this.authHttp.get('/api/auth/me');
    }
    return this.authMeCache;
  }

  private async supportsSubscriptionFamily(family: string, resource?: string): Promise<boolean> {
    try {
      const authMe = await this.authMe();
      if (!authMe?.has_active_subscription) {
        return false;
      }
      if (authMe.auth_type === 'user') {
        return true;
      }
      const capabilities = new Set(Array.isArray(authMe.capabilities) ? authMe.capabilities : []);
      if (capabilities.has(`${family}:*`)) {
        return true;
      }
      return Boolean(resource && capabilities.has(`${family}:${resource}`));
    } catch (error) {
      if (error instanceof APIError) {
        return false;
      }
      throw error;
    }
  }

  private async postFlow(flowType: string, payload: Record<string, any>): Promise<any> {
    const primary = (await this.supportsSubscriptionFamily('flows', flowType))
      ? `/agents/flow/${flowType}`
      : `/api/flow/${flowType}`;
    try {
      return await this.http.post(primary, payload);
    } catch (error) {
      if (primary.startsWith('/agents/flow/') && error instanceof APIError && (error.statusCode === 403 || error.statusCode === 404)) {
        return await this.http.post(`/api/flow/${flowType}`, payload);
      }
      throw error;
    }
  }

  private async getRender(renderId: string, status: boolean = false): Promise<any> {
    const suffix = status ? '/status' : '';
    const primary = (await this.supportsSubscriptionFamily('flows'))
      ? `/agents/flow/renders/${renderId}${suffix}`
      : `/api/flow/renders/${renderId}${suffix}`;
    try {
      return await this.http.get(primary);
    } catch (error) {
      if (primary.startsWith('/agents/flow/') && error instanceof APIError && (error.statusCode === 403 || error.statusCode === 404)) {
        return await this.http.get(`/api/flow/renders/${renderId}${suffix}`);
      }
      throw error;
    }
  }

  private async deleteRender(renderId: string): Promise<any> {
    const primary = (await this.supportsSubscriptionFamily('flows'))
      ? `/agents/flow/renders/${renderId}`
      : `/api/flow/renders/${renderId}`;
    try {
      return await this.http.delete(primary);
    } catch (error) {
      if (primary.startsWith('/agents/flow/') && error instanceof APIError && (error.statusCode === 403 || error.statusCode === 404)) {
        return await this.http.delete(`/api/flow/renders/${renderId}`);
      }
      throw error;
    }
  }

  /**
   * List all renders
   */
  async list(options?: {
    state?: string;
    template?: string;
    type?: string;
    tags?: string[];
  }): Promise<Render[]> {
    const params: Record<string, string> = {};
    if (options?.state) params.state = options.state;
    if (options?.template) params.template = options.template;
    if (options?.type) params.type = options.type;
    if (options?.tags?.length) {
      (params as unknown as Record<string, string | string[]>).tag = [...options.tags];
    }

    const data = await this.http.get('/api/renders', params);
    const items = typeof data === 'object' && data.items ? data.items : data;
    return (items || []).map(renderFromDict);
  }

  /**
   * Get render details
   */
  async get(renderId: string): Promise<Render> {
    const data = await this.getRender(renderId);
    return renderFromDict(data);
  }

  /**
   * Create a new render
   */
  async create(
    params: Record<string, any>,
    renderType: string = 'comfyui',
    notifyUrl?: string,
    tags?: string[],
  ): Promise<Render> {
    const payload: any = {
      type: renderType,
      params,
    };
    if (notifyUrl) {
      payload.notify_url = notifyUrl;
    }
    if (tags?.length) {
      payload.tags = [...tags];
    }

    const data = await this.http.post('/api/renders', payload);
    return renderFromDict(data);
  }

  /**
   * Cancel a render
   */
  async cancel(renderId: string): Promise<any> {
    return await this.deleteRender(renderId);
  }

  /**
   * Get render status (lightweight polling endpoint)
   */
  async status(renderId: string): Promise<RenderStatus> {
    const data = await this.getRender(renderId, true);
    return renderStatusFromDict(data);
  }

  async wait(
    renderId: string,
    options?: {
      timeoutMs?: number;
      pollIntervalMs?: number;
      queueGraceMs?: number;
      activeGraceMs?: number;
    },
  ): Promise<Render> {
    const timeoutMs = options?.timeoutMs ?? Renders.DEFAULT_WAIT_TIMEOUT;
    const pollIntervalMs = options?.pollIntervalMs ?? 5000;
    const queueGraceMs = options?.queueGraceMs ?? Renders.DEFAULT_QUEUE_GRACE;
    const activeGraceMs = options?.activeGraceMs ?? Renders.DEFAULT_ACTIVE_GRACE;

    let deadline = Date.now() + timeoutMs;
    let queueGraceUsed = false;
    let activeGraceUsed = false;
    let lastRender: Render | null = null;

    while (true) {
      const render = await this.get(renderId);
      lastRender = render;
      const state = (render.state || '').toLowerCase();
      if (state === 'completed' || state === 'failed' || state === 'cancelled') {
        return render;
      }

      const now = Date.now();
      if (now >= deadline) {
        const startedAt = this.parseRenderTimestamp(render.startedAt);
        if (!queueGraceUsed && startedAt == null) {
          queueGraceUsed = true;
          deadline = now + queueGraceMs;
        } else if (!activeGraceUsed && startedAt != null) {
          activeGraceUsed = true;
          deadline = Math.max(deadline, startedAt + activeGraceMs);
        } else {
          throw new Error(
            `Render ${renderId} did not complete within ${(timeoutMs / 1000).toFixed(0)}s ` +
            `(+${(queueGraceMs / 1000).toFixed(0)}s queue grace, +${(activeGraceMs / 1000).toFixed(0)}s active grace); ` +
            `lastRender=${JSON.stringify(lastRender)}`,
          );
        }
      }

      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }
  }

  // =========================================================================
  // Flow endpoints - simplified interfaces
  // =========================================================================

  private parseRenderTimestamp(value: unknown): number | null {
    if (value == null) return null;
    if (typeof value === 'number') return value;
    if (typeof value === 'string') {
      const parsed = Date.parse(value);
      return Number.isNaN(parsed) ? null : parsed;
    }
    return null;
  }

  async flow(flowType: string, params: Record<string, any>): Promise<Render> {
    // Filter out null/undefined values
    const payload: Record<string, any> = {};
    for (const [key, value] of Object.entries(params)) {
      if (value !== null && value !== undefined) {
        payload[key] = value;
      }
    }

    const data = await this.postFlow(flowType, payload);
    return renderFromDict(data);
  }

  /**
   * Generate an image using Qwen-Image
   */
  async textToImage(options: {
    prompt: string;
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('text-to-image', {
      prompt: options.prompt,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate an image using HiDream I1 Full
   */
  async textToImageHidream(options: {
    prompt: string;
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('text-to-image-hidream', {
      prompt: options.prompt,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate a video using Wan 2.2 14B
   */
  async textToVideo(options: {
    prompt: string;
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('text-to-video', {
      prompt: options.prompt,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Animate an image using Wan 2.2 Animate
   */
  async imageToVideo(options: {
    prompt: string;
    imageUrl?: string;
    fileIds?: string[];
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('image-to-video', {
      prompt: options.prompt,
      image_url: options.imageUrl,
      file_ids: options.fileIds,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate a lip-sync video using HuMo
   */
  async speakingVideo(options: {
    prompt: string;
    imageUrl?: string;
    audioUrl?: string;
    fileIds?: string[];
    negative?: string;
    length?: number;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('speaking-video', {
      prompt: options.prompt,
      image_url: options.imageUrl,
      audio_url: options.audioUrl,
      file_ids: options.fileIds,
      negative: options.negative,
      length: options.length,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate an audio-driven video using Wan 2.2 S2V
   */
  async speakingVideoWan(options: {
    prompt: string;
    imageUrl: string;
    audioUrl: string;
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('speaking-video-wan', {
      prompt: options.prompt,
      image_url: options.imageUrl,
      audio_url: options.audioUrl,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Transform images using Qwen Image Edit
   */
  async imageToImage(options: {
    prompt: string;
    imageUrls?: string[];
    fileIds?: string[];
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('image-to-image', {
      prompt: options.prompt,
      image_urls: options.imageUrls,
      file_ids: options.fileIds,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate video morphing between two images using Wan 2.2
   */
  async firstLastFrameVideo(options: {
    prompt: string;
    startImageUrl?: string;
    endImageUrl?: string;
    fileIds?: string[];
    negative?: string;
    width?: number;
    height?: number;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('first-last-frame-video', {
      prompt: options.prompt,
      start_image_url: options.startImageUrl,
      end_image_url: options.endImageUrl,
      file_ids: options.fileIds,
      negative: options.negative,
      width: options.width,
      height: options.height,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Transcribe audio/video to text using WhisperX
   */
  async audioToText(options: {
    audioUrl?: string;
    fileIds?: string[];
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('audio-to-text', {
      audio_url: options.audioUrl,
      file_ids: options.fileIds,
      notify_url: options.notifyUrl,
    });
  }

  /**
   * Generate speech from text using Qwen3-TTS
   */
  async textToSpeech(options: {
    text: string;
    mode?: string;
    language?: string;
    speaker?: string;
    style?: string;
    modelSize?: string;
    voiceDescription?: string;
    refAudioUrl?: string;
    fileIds?: string[];
    refText?: string;
    useXvectorOnly?: boolean;
    notifyUrl?: string;
  }): Promise<Render> {
    return this.flow('text-to-speech', {
      text: options.text,
      mode: options.mode,
      language: options.language,
      speaker: options.speaker,
      style: options.style,
      model_size: options.modelSize,
      voice_description: options.voiceDescription,
      ref_audio_url: options.refAudioUrl,
      file_ids: options.fileIds,
      ref_text: options.refText,
      use_xvector_only: options.useXvectorOnly,
      notify_url: options.notifyUrl,
    });
  }
}
